libkbfs/folder_block_manager.go

```diff
@@ -1365,12 +1435,40 @@ func (fbm *folderBlockManager) doCleanDiskCache(cacheType DiskBlockCacheType) (
 			return err
 		}
 
-		ptrs := getUnrefPointersFromMD(rmd)
-		ids := make([]kbfsblock.ID, len(ptrs))
-		for i, ptr := range ptrs {
-			ids[i] = ptr.ID
+		// Include unrefs from `gcOp`s here, as a double-check against
+		// archive races (see comment below).
+		ptrs := getUnrefPointersFromMD(rmd, true)
+
+		var ids []kbfsblock.ID
+		if cacheType == DiskBlockSyncCache {
+			// Wait for our own archives to complete, to make sure the
+			// bserver already knows this block isn't live yet when we
+			// make the call below.  However, when dealing with MDs
+			// written by other clients, there could be a race here
+			// where we see the ID is live before the other client
+			// gets to archive the block, leading to a leak.  Once the
+			// revision is GC'd though, we should run through this
+			// code again with the `gcOp`, and we'll delete the block
+			// then.  (Note there's always a chance for a race here,
+			// since the client could crash before archiving the
+			// blocks.  But the GC should always catch it eventually.)
+			err := fbm.waitForArchives(ctx)
+			if err != nil {
+				return err
+			}
+
+			ids, err = fbm.doChunkedGetNonLiveBlocks(ctx, ptrs)
```

Comment:
Comment that this blocks, but that's fine because this method is called in a background goroutine.